# Battle of the Neighborhoods - Coursera Capstone Project

**The best Location For an organization
** Introduction


This project will analyze neighborhoods between Toronto, Canada and New York City, New York. Organization is looking to move its headquarters to either Toronto or New York City. 
The company wants insight into the neighborhoods and local businesses in the cities so that its employees may have the optimum living standards and quality of life. 
This project will explore the similarities and dissimilarities between certain neighborhoods in the two cities, 
and determine which neighborhoods best fit the culture of the organization employees

# A description of the problem and a discussion of the background.

A Problem Is Here We Need To Understand The Similarities And Differences Of Neighbourhoods Between Scarborough Borough In Toronto And Queens Borough In New York City.
We Need To Select The Best Neighbourhood For An Organization To Move Its Headquarters Based On Venues

# A description of the data and how it will be used to solve the problem.


The data used for this project will be acquired from the respective cities Wikipedia website pages. The
datasets consists of the postal codes, neighborhood names, latitude, and longitude information for
each neighborhood. Foursquare API search feature will be used to collect neighborhood venue
information. Details about local venues and locality will be provide insight into the qualities of a
neighborhood. In addition to Foursquare, various python packages will be used to create maps and
machine learning models to further provide insights into our neighborhood battle project.
I used the following datasets from these websites:
Toronto Neighborhoods - https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M.
Toronto Latitude and Longitude - http://cocl.us/Geospatial_data
New York City neighborhoods - https://geo.nyu.edu/catalog/nyu_2451_34572
New York City Latitude and Longitude = Python Geolibrar

This Project Will Analyse Neighbourhoods Between Toronto, Canada And New York City, New York. For An Organization Is Looking To Move Its Headquarters To Either Toronto Or New York City. The Company Wants Insight Into The Neighbourhoods And Local Businesses In The Cities So That Its Employees May Have The Optimum Living Standards And Quality Of Life For Family And Himself. This Project Will Explore The Similarities And Dissimilarities Between Certain Neighbourhoods In The Two Cities, And Determine Which Neighbourhoods Best Fit The Culture Of The Organization For Employees.

# First we do Segmenting and Clustering Neighborhoods in Toronto City

**Use data from Toronto Neighborhood Wikipedia page to segment, cluster and explore neighborhoods in Toronto

In [3]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.19.0-py_0 conda-forge

geographiclib- 100% |################################| Time: 0:00:00  22.96 MB/s
geopy-1.19.0-p 100% |################################| Time: 0:00:00  11.49 MB/s
Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge

altair-2.2.2-p 100% |################################| Time: 0:00:00  10.76 MB/s
branca-0.3.1-p 100% |################################| Time: 0:00:00  29.44 MB/s
vincent-0.4.4- 100% |###################

**Download and Explore Toronto Dataset for analyses

The dateset being used is found at https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M.

The dataset is a list of Toronto's zipcodes which includes the boroughs and neighborhood names.

In [4]:
#Obtain Postal Code, Borough, and Neighborhood information from Wikipedia
table = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', header = 0)

#Obtain the first table
df_toronto = table[0]
df_toronto.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

**Transform the data for analyses

In [5]:
df_toronto.rename(columns = {"Postcode": "PostalCode", "Neighbourhood": "Neighborhood"}, inplace = True)

#Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
df_toronto.drop(df_toronto[df_toronto.Borough == 'Not assigned'].index, inplace=True)
#df.head()

#Combine the neighborhoods that exists in one postal code
df_toronto = df_toronto.groupby(['PostalCode', 'Borough'])['Neighborhood'].apply(lambda x: ','.join(x)).reset_index()
#df.head()

#Change unassigned Neighborhood to its Borough's name
df_toronto.loc[85,'Neighborhood'] = 'Queen\'s Park'

print (df_toronto.shape)

df_toronto.head()

(103, 3)


PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

**Join neighborhood table with latitude and longitude information

In [6]:
#Create a dataframe of the latitude and longitudes of the Toronto Neighborhoods
latlong = pd.read_csv("http://cocl.us/Geospatial_data")
latlong.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [8]:
latlong.rename(columns = {"Postal Code": "PostalCode"}, inplace = True)
latlong.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [9]:
latlong.tail()

PostalCode   Latitude  Longitude
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

**Here we Join latitude and longitude dataframe with neighborhood dataframe

In [10]:
#Join the Lat and Long dataframe to Neighborhoods dataframe
df_toronto.set_index("PostalCode")
latlong.set_index("PostalCode")
neighbor=pd.merge(df_toronto, latlong)
neighbor.head()

PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [11]:
print('Toronto has {} boroughs and {} neighborhoods.'.format(
        len(neighbor['Borough'].unique()),
        neighbor.shape[0]
    )
)

Toronto has 11 boroughs and 103 neighborhoods.


*Use geopy library to get the latitude and longitude values of Toronto, Canada.

In [12]:
address = 'Toronto, CA'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, Canada are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto, Canada are 43.653963, -79.387207.


/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.19.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


# Create a map of Toronto with neighborhoods superimposed on top.

In [13]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighbor['Latitude'], neighbor['Longitude'], neighbor['Borough'], neighbor['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

**Here let's simplify the above map and segment and cluster only the neighborhoods in Toronto. So let's slice the original dataframe and create a new dataframe of the Scarborough Neighborhood data.

In [14]:
scarborough_data = neighbor[neighbor['Borough'] == 'Scarborough'].reset_index(drop=True)
scarborough_data.head()

PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

**Lets get the geographical coordinates of Scarborough

In [15]:
address = 'Scarborough, Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Scarborough, CA are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Scarborough, CA are 43.773077, -79.257774.


/opt/conda/envs/DSX-Python35/lib/python3.5/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.19.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


# create map of Scarborough using latitude and longitude values

In [16]:
map_scarborough = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(scarborough_data['Latitude'], scarborough_data['Longitude'], scarborough_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_scarborough)  
    
map_scarborough

# Explore Scarborough neighborhood in Toronto with Foursquare API

**Define Foursquare credentials and version

In [17]:
CLIENT_ID = 'DMHXCRX0PYY5XT0H23OMWUCM5X15KRXOGC1KKDJUL0JNLGV3' # your Foursquare ID
CLIENT_SECRET = '3VTGVBA32HWX53O2E0XRIKWVLOPDLZWUCGDMCLDSB4V2FW4G' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: DMHXCRX0PYY5XT0H23OMWUCM5X15KRXOGC1KKDJUL0JNLGV3
CLIENT_SECRET:3VTGVBA32HWX53O2E0XRIKWVLOPDLZWUCGDMCLDSB4V2FW4G


# Here we explore the Function to explore neighborhoods Places and Venue.

In [21]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            100)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    print('Found {} venues in {} neighborhoods.'.format(nearby_venues.shape[0], len(venues_list)))
    
    return(nearby_venues)

In [22]:
scarborough_venues = getNearbyVenues(names=scarborough_data['Neighborhood'],
                                   latitudes=scarborough_data['Latitude'],
                                   longitudes=scarborough_data['Longitude']
                                  )

Found 88 venues in 17 neighborhoods.


In [23]:
print(scarborough_venues.shape)
scarborough_venues.head()

(88, 7)


Neighborhood  Neighborhood Latitude  \
0                         Rouge,Malvern              43.806686   
1  Highland Creek,Rouge Hill,Port Union              43.784535   
2  Highland Creek,Rouge Hill,Port Union              43.784535   
3  Highland Creek,Rouge Hill,Port Union              43.784535   
4       Guildwood,Morningside,West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497            Royal Canadian Legion       43.782533   
2              -79.160497        Affordable Toronto Movers       43.787919   
3              -79.160497   Scarborough Historical Society       43.788755   
4              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.162977         Moving Target  
3       -79.162438        History Museum  
4       -79.189914           Pizza Place

**Venues per Neighborhood

In [24]:
#Venues per Neighborhood
scarborough_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Agincourt North,L'Amoreaux East,Milliken,Steele...                      2   
Birch Cliff,Cliffside West                                              4   
Cedarbrae                                                               7   
Clairlea,Golden Mile,Oakridge                                          10   
Clarks Corners,Sullivan,Tam O'Shanter                                  10   
Cliffcrest,Cliffside,Scarborough Village West                           2   
Dorset Park,Scarborough Town Centre,Wexford Hei...                      7   
East Birchmount Park,Ionview,Kennedy Park                               6   
Guildwood,Morningside,West Hill                                         8   
Highland Creek,Rouge Hill,Port Union                                    3   
L'Amoreaux West                                                        13   
Maryvale,Wexford                                                        5   
Rouge,Malvern                                                           1   
Scarborough Village                                                     2   
Woburn                                                                  4   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Agincourt                                                                4   
Agincourt North,L'Amoreaux East,Milliken,Steele...                       2   
Birch Cliff,Cliffside West                                               4   
Cedarbrae                                                                7   
Clairlea,Golden Mile,Oakridge                                           10   
Clarks Corners,Sullivan,Tam O'Shanter                                   10   
Cliffcrest,Cliffside,Scarborough Village West                            2   
Dorset Park,Scarborough Town Centre,Wexford Hei...                       7   
East Birchmount Park,Ionview,Kennedy Park                                6   
Guildwood,Morningside,West Hill                                          8   
Highland Creek,Rouge Hill,Port Union                                     3   
L'Amoreaux West                                                         13   
Maryvale,Wexford                                                         5   
Rouge,Malvern                                                            1   
Scarborough Village                                                      2   
Woburn                                                                   4   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Agincourt                                               4               4   
Agincourt North,L'Amoreaux East,Milliken,Steele...      2               2   
Birch Cliff,Cliffside West                              4               4   
Cedarbrae                                               7               7   
Clairlea,Golden Mile,Oakridge                          10              10   
Clarks Corners,Sullivan,Tam O'Shanter                  10              10   
Cliffcrest,Cliffside,Scarborough Village West           2               2   
Dorset Park,Scarborough Town Centre,Wexford Hei...      7               7   
East Birchmount Park,Ionview,Kennedy Park               6               6   
Guildwood,Morningside,West Hill                         8               8   
Highland Creek,Rouge Hill,Port Union                    3               3   
L'Amoreaux West                                        13              13   
Maryvale,Wexford                                        5               5   
Rouge,Malvern                                           1               1   
Scarborough Village         

In [25]:
print('There are {} distinct venues in {} categories.'.format(
    len(scarborough_venues['Venue'].unique()),len(scarborough_venues['Venue Category'].unique())))

#print('There are {} uniques categories.'.format(len(scarborough_venues['Venue Category'].unique())))

There are 81 distinct venues in 54 categories.


# Now anlayse of each Neighborhood we do here

In [26]:
# one hot encoding
scarborough_onehot = pd.get_dummies(scarborough_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
scarborough_onehot['Neighborhood'] = scarborough_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [scarborough_onehot.columns[-1]] + list(scarborough_onehot.columns[:-1])
scarborough_onehot = scarborough_onehot[fixed_columns]

scarborough_onehot.head()

Neighborhood  American Restaurant  \
0                         Rouge,Malvern                    0   
1  Highland Creek,Rouge Hill,Port Union                    0   
2  Highland Creek,Rouge Hill,Port Union                    0   
3  Highland Creek,Rouge Hill,Port Union                    0   
4       Guildwood,Morningside,West Hill                    0   

   Athletics & Sports  Auto Garage  Bakery  Bank  Bar  Breakfast Spot  \
0                   0            0       0     0    0               0   
1                   0            0       0     0    1               0   
2                   0            0       0     0    0               0   
3                   0            0       0     0    0               0   
4                   0            0       0     0    0               0   

   Bus Line  Bus Station  Café  Camera Store  Caribbean Restaurant  \
0         0            0     0             0                     0   
1         0            0     0             0                     0   
2         0            0     0             0                     0   
3         0            0     0             0                     0   
4         0            0     0             0                     0   

   Chinese Restaurant  Clothing Store  Coffee Shop  College Stadium  \
0                   0               0            0                0   
1                   0               0            0                0   
2                   0               0            0                0   
3                   0               0            0                0   
4                   0               0            0                0   

   Convenience Store  Department Store  Discount Store  Electronics Store  \
0                  0                 0               0                  0   
1                  0                 0               0                  0   
2                  0                 0               0                  0   
3                  0                 0               0                  0   
4                  0                 0               0                  0   

   Fast Food Restaurant  Fried Chicken Joint  Furniture / Home Store  \
0                     1                    0                       0   
1                     0                    0                       0   
2                     0                    0                       0   
3                     0                    0                       0   
4                     0                    0                       0   

   General Entertainment  Grocery Store  Gym Pool  Hakka Restaurant  \
0                      0              0         0                 0   
1                      0              0         0                 0   
2                      0              0         0                 0   
3                      0              0         0                 0   
4                      0              0         0                 0   

   History Museum  Indian Restaurant  Insurance Office  Intersection  \
0               0                  0                 0             0   
1               0                  0                 0             0   
2               0                  0                 0             0   
3               1                  0                 0             0   
4               0                  0                 0             0   

   Italian Restaurant  Korean Restaurant  Latin American Restaurant  Lounge  \
0                   0                  0                          0       0   
1                   0                  0                          0       0   
2                   0                  0                          0       0   
3                   0                  0                          0       0   
4                   0                  0                          0       0   

   Medical Center  Metro Station  Mexican Restaurant  \
0               0              0                   0   
1           

**let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category in the table

In [27]:
scarborough_grouped = scarborough_onehot.groupby('Neighborhood').mean().reset_index()
scarborough_grouped

Neighborhood  American Restaurant  \
0                                           Agincourt                  0.0   
1   Agincourt North,L'Amoreaux East,Milliken,Steel...                  0.0   
2                          Birch Cliff,Cliffside West                  0.0   
3                                           Cedarbrae                  0.0   
4                       Clairlea,Golden Mile,Oakridge                  0.0   
5               Clarks Corners,Sullivan,Tam O'Shanter                  0.0   
6       Cliffcrest,Cliffside,Scarborough Village West                  0.5   
7   Dorset Park,Scarborough Town Centre,Wexford He...                  0.0   
8           East Birchmount Park,Ionview,Kennedy Park                  0.0   
9                     Guildwood,Morningside,West Hill                  0.0   
10               Highland Creek,Rouge Hill,Port Union                  0.0   
11                                    L'Amoreaux West                  0.0   
12                                   Maryvale,Wexford                  0.0   
13                                      Rouge,Malvern                  0.0   
14                                Scarborough Village                  0.0   
15                                             Woburn                  0.0   

    Athletics & Sports  Auto Garage    Bakery      Bank       Bar  \
0             0.000000          0.0  0.000000  0.000000  0.000000   
1             0.000000          0.0  0.000000  0.000000  0.000000   
2             0.000000          0.0  0.000000  0.000000  0.000000   
3             0.142857          0.0  0.142857  0.142857  0.000000   
4             0.000000          0.0  0.200000  0.000000  0.000000   
5             0.000000          0.0  0.000000  0.000000  0.000000   
6             0.000000          0.0  0.000000  0.000000  0.000000   
7             0.000000          0.0  0.000000  0.000000  0.000000   
8             0.000000          0.0  0.000000  0.000000  0.000000   
9             0.000000          0.0  0.000000  0.000000  0.000000   
10            0.000000          0.0  0.000000  0.000000  0.333333   
11            0.000000          0.0  0.000000  0.000000  0.000000   
12            0.000000          0.2  0.200000  0.000000  0.000000   
13            0.000000          0.0  0.000000  0.000000  0.000000   
14            0.000000          0.0  0.000000  0.000000  0.000000   
15            0.000000          0.0  0.000000  0.000000  0.000000   

    Breakfast Spot  Bus Line  Bus Station  Café  Camera Store  \
0         0.250000       0.0     0.000000  0.00      0.000000   
1         0.000000       0.0     0.000000  0.00      0.000000   
2         0.000000       0.0     0.000000  0.25      0.000000   
3         0.000000       0.0     0.000000  0.00      0.000000   
4         0.000000       0.2     0.100000  0.00      0.000000   
5         0.000000       0.0     0.000000  0.00      0.000000   
6         0.000000       0.0     0.000000  0.00      0.000000   
7         0.000000       0.0     0.000000  0.00      0.000000   
8         0.000000       0.0     0.166667  0.00      0.000000   
9         0.125000       0.0     0.000000  0.00      0.000000   
10        0.000000       0.0     0.000000  0.00      0.000000   
11        0.076923       0.0     0.000000  0.00      0.076923   
12        0.200000       0.0     0.000000  0.00      0.000000   
13        0.000000       0.0     0.000000  0.00      0.000000   
14        0.000000       0.0     0.000000  0.00      0.000000   
15        0.000000       0.0     0.000000  0.00      0.000000   

    Caribbean Restaurant  Chinese Restaurant  Clothing Store  Coffee Shop  \
0               0.000000            0.000000            0.25     0.000000   
1               0.000000            0.000000            0.00     0.000000   
2               0.000000            0.000000            0.00     0.000000   
3               0.142857            0.000000            0.00     0.000000   
4               0.000000            0.000000     

# Here we get Each neighborhood with the top 5 venues.

In [28]:
num_top_venues = 10

for hood in scarborough_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = scarborough_grouped[scarborough_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0               Skating Rink  0.25
1             Breakfast Spot  0.25
2                     Lounge  0.25
3             Clothing Store  0.25
4        American Restaurant  0.00
5              Moving Target  0.00
6               Intersection  0.00
7         Italian Restaurant  0.00
8          Korean Restaurant  0.00
9  Latin American Restaurant  0.00


----Agincourt North,L'Amoreaux East,Milliken,Steeles East----
                       venue  freq
0                 Playground   0.5
1                       Park   0.5
2        American Restaurant   0.0
3                      Motel   0.0
4           Insurance Office   0.0
5               Intersection   0.0
6         Italian Restaurant   0.0
7          Korean Restaurant   0.0
8  Latin American Restaurant   0.0
9                     Lounge   0.0


----Birch Cliff,Cliffside West----
                       venue  freq
0            College Stadium  0.25
1               Skating Rink  0.25
2     

**Let's put data into a pandas dataframe

In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

**Create a new dataframe and display the top ten venues for each neighborhood

In [30]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = scarborough_grouped['Neighborhood']

for ind in np.arange(scarborough_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(scarborough_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood  1st Most Common Venue  \
0                                           Agincourt         Clothing Store   
1   Agincourt North,L'Amoreaux East,Milliken,Steel...             Playground   
2                          Birch Cliff,Cliffside West  General Entertainment   
3                                           Cedarbrae       Hakka Restaurant   
4                       Clairlea,Golden Mile,Oakridge                 Bakery   
5               Clarks Corners,Sullivan,Tam O'Shanter            Pizza Place   
6       Cliffcrest,Cliffside,Scarborough Village West    American Restaurant   
7   Dorset Park,Scarborough Town Centre,Wexford He...      Indian Restaurant   
8           East Birchmount Park,Ionview,Kennedy Park     Chinese Restaurant   
9                     Guildwood,Morningside,West Hill            Pizza Place   
10               Highland Creek,Rouge Hill,Port Union         History Museum   
11                                    L'Amoreaux West   Fast Food Restaurant   
12                                   Maryvale,Wexford            Auto Garage   
13                                      Rouge,Malvern   Fast Food Restaurant   
14                                Scarborough Village                    Spa   
15                                             Woburn            Coffee Shop   

    2nd Most Common Venue  3rd Most Common Venue   4th Most Common Venue  \
0            Skating Rink         Breakfast Spot                  Lounge   
1                    Park  Vietnamese Restaurant      Chinese Restaurant   
2            Skating Rink                   Café         College Stadium   
3      Athletics & Sports                 Bakery                    Bank   
4                Bus Line           Intersection            Soccer Field   
5            Noodle House    Fried Chicken Joint         Thai Restaurant   
6                   Motel  General Entertainment  Furniture / Home Store   
7   Vietnamese Restaurant              Pet Store  Furniture / Home Store   
8          Discount Store            Bus Station        Department Store   
9     Rental Car Location           Intersection          Breakfast Spot   
10                    Bar          Moving Target   Vietnamese Restaurant   
11     Chinese Restaurant           Camera Store                Pharmacy   
12                 Bakery         Sandwich Place          Breakfast Spot   
13  Vietnamese Restaurant         Clothing Store           Grocery Store   
14             Playground  Vietnamese Restaurant      Chinese Restaurant   
15       Insurance Office      Korean Restaurant   Vietnamese Restaurant   

        5th Most Common Venue   6th Most Common Venue   7th Most Common Venue  \
0       Vietnamese Restaurant   General Entertainment  Furniture / Home Store   
1       General Entertainment  Furniture / Home Store     Fried Chicken Joint   
2       Vietnamese Restaurant          Clothing Store  Furniture / Home Store   
3         Fried Chicken Joint    Caribbean Restaurant         Thai Restaurant   
4        Fast Food Restaurant           Metro Station             Bus Station   
5        Fast Food Restaurant     Rental Car Location      Chinese Restaurant   
6         Fried Chicken Joint    Fast Food Restaurant       Electronics Store   
7   Latin American Restaurant      Chinese Restaurant                     Bar   
8           Convenience Store             Coffee Shop          Clothing Store   
9              Medical Center      Mexican Restaurant       Electronics Store   
10             Clothing Store   General Entertainment  Furniture / Home Store   
11             Discount Store             Coffee Shop           Grocery Store   
12  Middle Eastern Restaurant   Vietnamese Restaurant             Coffee Shop   
13      General Entertainment  Furniture / Home Store     Fried Chicken Joint   
14      General Entertainment  Furniture / Home Store     Fried Chicken Joint   
15                   Gym Pool   General Entertainment  Furniture / Home Store   

    8th M

In [25]:
neighborhoods_venues_sorted.iloc[11,]

Neighborhood                   L'Amoreaux West
1st Most Common Venue     Fast Food Restaurant
2nd Most Common Venue       Chinese Restaurant
3rd Most Common Venue             Camera Store
4th Most Common Venue                 Pharmacy
5th Most Common Venue           Discount Store
6th Most Common Venue              Coffee Shop
7th Most Common Venue            Grocery Store
8th Most Common Venue           Breakfast Spot
9th Most Common Venue                 Gym Pool
10th Most Common Venue             Pizza Place
Name: 11, dtype: object

 # Cluster the Scarborough Neighborhood using k-means¶

**Run K-means to cluster neighborhood into three clusters

In [31]:
# set number of clusters
kclusters = 3

scarborough_grouped_clustering = scarborough_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=2).fit(scarborough_grouped_clustering)

# check cluster labels generated for each row in the dataframe
#kmeans.labels_[0:10] 
kmeans.labels_

array([1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 0, 1], dtype=int32)

**Lets create a new dataframe that includes the cluster as well as the top ten venues for each neighborhood

In [32]:
#Note that the neighborhood Upper Rouge does not have any venues, so I will drop from dataset
scarborough_data.drop(scarborough_data[scarborough_data.Neighborhood == 'Upper Rouge'].index, inplace = True)
#df_toronto.drop(df_toronto[df_toronto.Borough == 'Not assigned'].index, inplace=True)

scarborough_merged = scarborough_data

# add clustering labels
scarborough_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
scarborough_merged = scarborough_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

scarborough_merged.head() # check the last columns!

PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0 -79.194353               1  Fast Food Restaurant  Vietnamese Restaurant   
1 -79.160497               0        History Museum                    Bar   
2 -79.188711               1           Pizza Place    Rental Car Location   
3 -79.216917               1           Coffee Shop       Insurance Office   
4 -79.239476               1      Hakka Restaurant     Athletics & Sports   

  3rd Most Common Venue  4th Most Common Venue  5th Most Common Venue  \
0        Clothing Store          Grocery Store  General Entertainment   
1         Moving Target  Vietnamese Restaurant         Clothing Store   
2          Intersection         Breakfast Spot         Medical Center   
3     Korean Restaurant  Vietnamese Restaurant               Gym Pool   
4                Bakery                   Bank    Fried Chicken Joint   

    6th Most Common Venue   7th Most Common Venue 8th Most Common Venue  \
0  Furniture / Home Store     Fried Chicken Joint     Electronics Store   
1   General Entertainment  Furniture / Home Store   Fried Chicken Joint   
2      Mexican Restaurant       Electronics Store                   Spa   
3   General Entertainment  Furniture / Home Store   Fried Chicken Joint   
4    Caribbean Restaurant         Thai Restaurant       College Stadium   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store       Department Store  
1  Fast Food Restaurant      Electronics Store  
2        Discount Store   Fast Food Restaurant  
3  Fast Food Restaurant      Electronics Store  
4         Grocery Store  General Entertainment

In [33]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(scarborough_merged['Latitude'], scarborough_merged['Longitude'], scarborough_merged['Neighborhood'], scarborough_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examine Scarborough Neighborhood Cluster


**Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster.

**Scarborough Clusters 0, 1, 2

In [34]:
scarborough_cluster_0 = scarborough_merged.loc[scarborough_merged['Cluster Labels'] == 0, scarborough_merged.columns[[1] + list(range(4, scarborough_merged.shape[1]))]]

scarborough_cluster_1 = scarborough_merged.loc[scarborough_merged['Cluster Labels'] == 1, scarborough_merged.columns[[1] + list(range(4, scarborough_merged.shape[1]))]]

scarborough_cluster_2 = scarborough_merged.loc[scarborough_merged['Cluster Labels'] == 2, scarborough_merged.columns[[1] + list(range(4, scarborough_merged.shape[1]))]]

In [35]:
scarborough_cluster_0

Borough  Longitude  Cluster Labels 1st Most Common Venue  \
1   Scarborough -79.160497               0        History Museum   
14  Scarborough -79.284577               0            Playground   

   2nd Most Common Venue  3rd Most Common Venue  4th Most Common Venue  \
1                    Bar          Moving Target  Vietnamese Restaurant   
14                  Park  Vietnamese Restaurant     Chinese Restaurant   

    5th Most Common Venue   6th Most Common Venue   7th Most Common Venue  \
1          Clothing Store   General Entertainment  Furniture / Home Store   
14  General Entertainment  Furniture / Home Store     Fried Chicken Joint   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1    Fried Chicken Joint  Fast Food Restaurant      Electronics Store  
14  Fast Food Restaurant     Electronics Store         Discount Store

In [32]:
scarborough_cluster_1

Borough  Longitude  Cluster Labels  1st Most Common Venue  \
0   Scarborough -79.194353               1   Fast Food Restaurant   
2   Scarborough -79.188711               1            Pizza Place   
3   Scarborough -79.216917               1            Coffee Shop   
4   Scarborough -79.239476               1       Hakka Restaurant   
5   Scarborough -79.239476               1                    Spa   
6   Scarborough -79.262029               1     Chinese Restaurant   
7   Scarborough -79.284577               1                 Bakery   
8   Scarborough -79.239476               1    American Restaurant   
9   Scarborough -79.264848               1  General Entertainment   
10  Scarborough -79.273304               1      Indian Restaurant   
11  Scarborough -79.295849               1            Auto Garage   
12  Scarborough -79.262029               1         Clothing Store   
15  Scarborough -79.318389               1   Fast Food Restaurant   

    2nd Most Common Venue  3rd Most Common Venue   4th Most Common Venue  \
0   Vietnamese Restaurant         Clothing Store           Grocery Store   
2     Rental Car Location           Intersection          Breakfast Spot   
3        Insurance Office      Korean Restaurant   Vietnamese Restaurant   
4      Athletics & Sports                 Bakery                    Bank   
5              Playground  Vietnamese Restaurant      Chinese Restaurant   
6          Discount Store            Bus Station        Department Store   
7                Bus Line           Intersection            Soccer Field   
8                   Motel  General Entertainment  Furniture / Home Store   
9            Skating Rink                   Café         College Stadium   
10  Vietnamese Restaurant              Pet Store  Furniture / Home Store   
11                 Bakery         Sandwich Place          Breakfast Spot   
12           Skating Rink         Breakfast Spot                  Lounge   
15     Chinese Restaurant           Camera Store                Pharmacy   

        5th Most Common Venue   6th Most Common Venue   7th Most Common Venue  \
0       General Entertainment  Furniture / Home Store     Fried Chicken Joint   
2              Medical Center      Mexican Restaurant       Electronics Store   
3                    Gym Pool   General Entertainment  Furniture / Home Store   
4         Fried Chicken Joint    Caribbean Restaurant         Thai Restaurant   
5       General Entertainment  Furniture / Home Store     Fried Chicken Joint   
6           Convenience Store             Coffee Shop          Clothing Store   
7        Fast Food Restaurant           Metro Station             Bus Station   
8         Fried Chicken Joint    Fast Food Restaurant       Electronics Store   
9       Vietnamese Restaurant          Clothing Store  Furniture / Home Store   
10  Latin American Restaurant      Chinese Restaurant                     Bar   
11  Middle Eastern Restaurant   Vietnamese Restaurant             Coffee Shop   
12      Vietnamese Restaurant   General Entertainment  Furniture / Home Store   
15             Discount Store             Coffee Shop           Grocery Store   

    8th Most Common Venue   9th Most Common Venue  10th Most Common Venue  
0       Electronics Store          Discount Store        Department Store  
2                     Spa          Discount Store    Fast Food Restaurant  
3     Fried Chicken Joint    Fast Food Restaurant       Electronics Store  
4         College Stadium           Grocery Store   General Entertainment  
5    Fast Food Restaurant       Electronics Store          Discount Store  
6           Grocery Store   General Entertainment  Furniture / Home Store  
7                    Park          Discount Store         College Stadium  
8          Discount Store        Department Store       Convenience Store  
9     Fried Chicken Joint    Fast Food Restaurant       Electronics Store  
10         Breakfast Spot      Athletics & Sports             Auto Garage  
11  G

In [36]:
scarborough_cluster_2

Borough  Longitude  Cluster Labels 1st Most Common Venue  \
13  Scarborough -79.304302               2           Pizza Place   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13          Noodle House   Fried Chicken Joint       Thai Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13  Fast Food Restaurant   Rental Car Location    Chinese Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13              Pharmacy    Italian Restaurant            Coffee Shop

# Explore New York City Neighborhoods

New York City has a total of 5 boroughs and 306 neighborhoods. In order to segement the neighborhoods and explore them, we will essentially need a dataset that contains the 5 boroughs and the neighborhoods that exist in each borough as well as the the latitude and logitude coordinates of each neighborhood.

Luckily, this dataset exists for free on the web. The link to the dataset: https://geo.nyu.edu/catalog/nyu_2451_34572

For your convenience, I downloaded the files and placed it on the server, so you can simply run a wget command and access the data. So let's go ahead and do that.

In [37]:
!wget -q -O 'newyork_data.json' https://ibm.box.com/shared/static/fbpwbovar7lf8p5sgddm06cgipa2rxpe.json
print('Data downloaded!')

Data downloaded!


**Load and explore dataset we have.

In [38]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

Notice how all the relevant data is in the features key, which is basically a list of the neighborhoods. So, let's define a new variable that includes this data.

In [39]:
neighborhoods_data = newyork_data['features']

In [38]:
neighborhoods_data[0]

{'geometry': {'coordinates': [-73.84720052054902, 40.89470517661],
  'type': 'Point'},
 'geometry_name': 'geom',
 'id': 'nyu_2451_34572.1',
 'properties': {'annoangle': 0.0,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661],
  'borough': 'Bronx',
  'name': 'Wakefield',
  'stacked': 1},
 'type': 'Feature'}

# Transform the data into a pandas dataframe

In [40]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

In [41]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)


In [41]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [42]:

print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


# Use Geolibrary to get the latitude and longitude of New York City

In [43]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent = 'my-application')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7308619, -73.9871558.


In [45]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork


**Lets simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Queens neighborhood data.

In [46]:
queens_data = neighborhoods[neighborhoods['Borough'] == 'Queens'].reset_index(drop=True)
queens_data.head()

Borough     Neighborhood   Latitude  Longitude
0  Queens          Astoria  40.768509 -73.915654
1  Queens         Woodside  40.746349 -73.901842
2  Queens  Jackson Heights  40.751981 -73.882821
3  Queens         Elmhurst  40.744049 -73.881656
4  Queens     Howard Beach  40.654225 -73.838138

Lets get the geographical location of Queens, NY

In [47]:

address = 'Queens, NY'

geolocator = Nominatim(user_agent = 'my-application')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Queens are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Queens are 40.6524927, -73.7914214158161.


In [48]:
# create map of Manhattan using latitude and longitude values
map_queens = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(queens_data['Latitude'], queens_data['Longitude'], queens_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_queens)  
    
map_queens

# Lets explore the Long Island City neighborhood in Queens, NY

In [49]:

queens_data.loc[10, 'Neighborhood']

'Long Island City'

In [50]:

#Long Island City Latitude and Longitude values

neighborhood_latitude = queens_data.loc[10, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = queens_data.loc[10, 'Longitude'] # neighborhood longitude value

neighborhood_name = queens_data.loc[10, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Long Island City are 40.75021734610528, -73.93920223915505.


# Top 100 venues in Long Island City neighborhood within a radius of 500 meters

First, let's create the GET request URL named url.

In [51]:
# type your answer here
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

url

'https://api.foursquare.com/v2/venues/explore?&client_id=DMHXCRX0PYY5XT0H23OMWUCM5X15KRXOGC1KKDJUL0JNLGV3&client_secret=3VTGVBA32HWX53O2E0XRIKWVLOPDLZWUCGDMCLDSB4V2FW4G&v=20180605&ll=40.75021734610528,-73.93920223915505&radius=500&limit=100'

In [52]:
#Send the GET request
results = requests.get(url).json()

In [53]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [54]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name          categories  \
0  Hilton Garden Inn New York Long Island City/Ma...               Hotel   
1                                 Baker House Market   Convenience Store   
2                                   Clever Blend Lic         Coffee Shop   
3                                  Etto Espresso Bar         Coffee Shop   
4                                             Waheyo  Mexican Restaurant   

         lat        lng  
0  40.750216 -73.936886  
1  40.752137 -73.939235  
2  40.750228 -73.939608  
3  40.748703 -73.940689  
4  40.750016 -73.939360

In [55]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

69 venues were returned by Foursquare.


# Analyze Each Neighborhood in Queens

In [56]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    print('Found {} venues in {} neighborhoods.'.format(nearby_venues.shape[0], len(venues_list)))
    
    return(nearby_venues)

In [58]:
queens_venues = getNearbyVenues(names=queens_data['Neighborhood'],
                                   latitudes=queens_data['Latitude'],
                                   longitudes=queens_data['Longitude'])

Found 2160 venues in 81 neighborhoods.


In [59]:
queens_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0      Astoria              40.768509              -73.915654   
1      Astoria              40.768509              -73.915654   
2      Astoria              40.768509              -73.915654   
3      Astoria              40.768509              -73.915654   
4      Astoria              40.768509              -73.915654   

                Venue  Venue Latitude  Venue Longitude        Venue Category  
0        Favela Grill       40.767348       -73.917897  Brazilian Restaurant  
1      Orange Blossom       40.769856       -73.917012          Gourmet Shop  
2    Titan Foods Inc.       40.769198       -73.919253          Gourmet Shop  
3     CrossFit Queens       40.769404       -73.918977                   Gym  
4  Simply Fit Astoria       40.769114       -73.912403                   Gym

In [60]:
print(queens_venues.shape)
queens_venues.tail()

(2160, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
2155  Queensbridge              40.756091              -73.945631   
2156  Queensbridge              40.756091              -73.945631   
2157  Queensbridge              40.756091              -73.945631   
2158  Queensbridge              40.756091              -73.945631   
2159  Queensbridge              40.756091              -73.945631   

                                  Venue  Venue Latitude  Venue Longitude  \
2155  Queensbridge Park Softball Fields       40.756055       -73.948407   
2156    SummerStage - Queensbridge Park       40.755514       -73.949125   
2157                The Ravel Hotel Gym       40.753787       -73.948815   
2158                          Track 114       40.753008       -73.947833   
2159                      Ravel Rooftop       40.753971       -73.949456   

             Venue Category  
2155         Baseball Field  
2156  Performing Arts Venue  
2157     Athletics & Sports  
2158               Platform  
2159              Roof Deck

In [59]:

#Venues per Neighborhood
queens_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                 
Arverne                                  17                      17     17   
Astoria                                 100                     100    100   
Astoria Heights                          12                      12     12   
Auburndale                               19                      19     19   
Bay Terrace                              42                      42     42   
Bayside                                  71                      71     71   
Bayswater                                 4                       4      4   
Beechhurst                               15                      15     15   
Bellaire                                 12                      12     12   
Belle Harbor                             16                      16     16   
Bellerose                                22                      22     22   
Blissville                               22                      22     22   
Breezy Point                              6                       6      6   
Briarwood                                14                      14     14   
Broad Channel                             4                       4      4   
Brookville                                2                       2      2   
Cambria Heights                          14                      14     14   
College Point                            44                      44     44   
Corona                                   17                      17     17   
Douglaston                               19                      19     19   
East Elmhurst                            17                      17     17   
Edgemere                                 14                      14     14   
Elmhurst                                 37                      37     37   
Far Rockaway                             35                      35     35   
Floral Park                               8                       8      8   
Flushing                                 69                      69     69   
Forest Hills                             37                      37     37   
Forest Hills Gardens                     23                      23     23   
Fresh Meadows                            13                      13     13   
Glen Oaks                                18                      18     18   
Glendale                                  4                       4      4   
Hammels                                  18                      18     18   
Hillcrest                                17                      17     17   
Hollis                                   18                      18     18   
Holliswood                                4                       4      4   
Howard Beach                             42                      42     42   
Hunters Point                            76                      76     76   
Jackson Heights                          84                      84     84   
Jamaica Center                           41                      41     41   
Jamaica Estates                           4                       4      4   
Jamaica Hills                            26                      26     26   
Kew Gardens                              44                      44     44   
Kew Gardens Hills                        21                      21     21   
Laurelton                                 6                       6      6   
Lefrak City                              27                      27     27   
Lindenwood                               13                      13     13   
Little Neck                              52                      52     52   
Long Island City                         69                      69     69   
Malba                                     3                       3      3   
Maspeth                                  31 

In [61]:
print('There are {} distinct venues in {} categories.'.format(
    len(queens_venues['Venue'].unique()),len(queens_venues['Venue Category'].unique())))

There are 1805 distinct venues in 269 categories.


In [64]:
# one hot encoding
queens_onehot = pd.get_dummies(queens_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
queens_onehot['Neighborhood'] = queens_venues['Neighborhood'] 

# move neighborhood column to the first column
#fixed_columns = [scarborough_onehot.columns[-1]] + list(scarborough_onehot.columns[:-1])
#scarborough_onehot = scarborough_onehot[fixed_columns]

neighbor = queens_onehot['Neighborhood']
queens_onehot.drop(labels=['Neighborhood'], axis=1,inplace = True)
queens_onehot.insert(0, 'Neighborhood', neighbor)

queens_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  Airport Terminal  \
0      Astoria                  0                  0                 0   
1      Astoria                  0                  0                 0   
2      Astoria                  0                  0                 0   
3      Astoria                  0                  0                 0   
4      Astoria                  0                  0                 0   

   American Restaurant  Arepa Restaurant  Argentinian Restaurant  Art Gallery  \
0                    0                 0                       0            0   
1                    0                 0                       0            0   
2                    0                 0                       0            0   
3                    0                 0                       0            0   
4                    0                 0                       0            0   

   Art Museum  Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0           0                    0                     0                 0   
1           0                    0                     0                 0   
2           0                    0                     0                 0   
3           0                    0                     0                 0   
4           0                    0                     0                 0   

   Athletics & Sports  Automotive Shop  BBQ Joint  Bagel Shop  Bakery  Bank  \
0                   0                0          0           0       0     0   
1                   0                0          0           0       0     0   
2                   0                0          0           0       0     0   
3                   0                0          0           0       0     0   
4                   0                0          0           0       0     0   

   Bar  Baseball Field  Basketball Court  Bath House  Beach  Beach Bar  \
0    0               0                 0           0      0          0   
1    0               0                 0           0      0          0   
2    0               0                 0           0      0          0   
3    0               0                 0           0      0          0   
4    0               0                 0           0      0          0   

   Bed & Breakfast  Beer Garden  Bistro  Board Shop  Bookstore  Boutique  \
0                0            0       0           0          0         0   
1                0            0       0           0          0         0   
2                0            0       0           0          0         0   
3                0            0       0           0          0         0   
4                0            0       0           0          0         0   

   Bowling Alley  Brazilian Restaurant  Breakfast Spot  Brewery  \
0              0                     1               0        0   
1              0                     0               0        0   
2              0                     0               0        0   
3              0                     0               0        0   
4              0                     0               0        0   

   Bubble Tea Shop  Buffet  Building  Burger Joint  Burrito Place  \
0                0       0         0             0              0   
1                0       0         0             0              0   
2                0       0         0             0              0   
3                0       0         0             0              0   
4                0       0         0             0              0   

   Bus Station  Bus Stop  Business Service  Café  Cajun / Creole Restaurant  \
0            0         0                 0     0                          0   
1            0         0                 0     0                          0   
2            0         0                 0     0                          0   
3            0         0                 0     0                          0   
4          

# Group by Neighborhood, and fine the frequency of the occurrence of venue

In [65]:
queens_grouped = queens_onehot.groupby('Neighborhood').mean().reset_index()
queens_grouped

Neighborhood  Accessories Store  Afghan Restaurant  \
0                Arverne           0.000000           0.000000   
1                Astoria           0.000000           0.000000   
2        Astoria Heights           0.000000           0.000000   
3             Auburndale           0.000000           0.000000   
4            Bay Terrace           0.023810           0.000000   
5                Bayside           0.000000           0.000000   
6              Bayswater           0.000000           0.000000   
7             Beechhurst           0.000000           0.000000   
8               Bellaire           0.000000           0.000000   
9           Belle Harbor           0.000000           0.000000   
10             Bellerose           0.000000           0.000000   
11            Blissville           0.000000           0.000000   
12          Breezy Point           0.000000           0.000000   
13             Briarwood           0.000000           0.000000   
14         Broad Channel           0.000000           0.000000   
15            Brookville           0.000000           0.000000   
16       Cambria Heights           0.000000           0.000000   
17         College Point           0.000000           0.000000   
18                Corona           0.000000           0.000000   
19            Douglaston           0.000000           0.000000   
20         East Elmhurst           0.000000           0.000000   
21              Edgemere           0.000000           0.000000   
22              Elmhurst           0.000000           0.000000   
23          Far Rockaway           0.000000           0.000000   
24           Floral Park           0.000000           0.000000   
25              Flushing           0.000000           0.000000   
26          Forest Hills           0.000000           0.000000   
27  Forest Hills Gardens           0.000000           0.000000   
28         Fresh Meadows           0.000000           0.000000   
29             Glen Oaks           0.000000           0.000000   
30              Glendale           0.000000           0.000000   
31               Hammels           0.000000           0.000000   
32             Hillcrest           0.000000           0.000000   
33                Hollis           0.000000           0.000000   
34            Holliswood           0.000000           0.000000   
35          Howard Beach           0.000000           0.000000   
36         Hunters Point           0.000000           0.000000   
37       Jackson Heights           0.000000           0.000000   
38        Jamaica Center           0.000000           0.000000   
39       Jamaica Estates           0.000000           0.000000   
40         Jamaica Hills           0.000000           0.000000   
41           Kew Gardens           0.000000           0.000000   
42     Kew Gardens Hills           0.000000           0.000000   
43             Laurelton           0.000000           0.000000   
44           Lefrak City           0.000000           0.000000   
45            Lindenwood           0.000000           0.000000   
46           Little Neck           0.000000           0.000000   
47      Long Island City           0.000000           0.000000   
48                 Malba           0.000000           0.000000   
49               Maspeth           0.000000           0.000000   
50        Middle Village           0.000000           0.000000   
51           Murray Hill           0.000000           0.000000   
52              Neponsit           0.000000           0.000000   
53          North Corona           0.000000           0.000000   
54       Oakland Gardens           0.000000           0.000000   
55            Ozone Park           0.000000           0.000000   
56               Pomonok           0.000000           0.000000   
57        Queens Village           0.000000           0.000000   
58       Queensboro Hill           0.000000           0.000000   
59          Queensbridge           0.000000         

# Here we find Each Neighborhood with the top 5 venues

In [66]:

num_top_venues = 5

for hood in queens_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = queens_grouped[queens_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Arverne----
             venue  freq
0        Surf Spot  0.24
1    Metro Station  0.12
2  Thai Restaurant  0.06
3      Pizza Place  0.06
4       Board Shop  0.06


----Astoria----
                       venue  freq
0                        Bar  0.07
1  Middle Eastern Restaurant  0.06
2                 Hookah Bar  0.06
3           Greek Restaurant  0.05
4         Seafood Restaurant  0.04


----Astoria Heights----
                venue  freq
0         Pizza Place  0.08
1        Burger Joint  0.08
2               Motel  0.08
3  Italian Restaurant  0.08
4       Moving Target  0.08


----Auburndale----
                    venue  freq
0      Italian Restaurant  0.11
1           Deli / Bodega  0.05
2              Hookah Bar  0.05
3  Furniture / Home Store  0.05
4       Mobile Phone Shop  0.05


----Bay Terrace----
                 venue  freq
0       Clothing Store  0.12
1        Women's Store  0.07
2           Donut Shop  0.05
3  American Restaurant  0.05
4       Lingerie Store  0.05


-

                venue  freq
0  Chinese Restaurant  0.09
1            Pharmacy  0.05
2          Donut Shop  0.05
3                 Bar  0.05
4                Bank  0.05


----Kew Gardens Hills----
              venue  freq
0       Bus Station  0.14
1       Pizza Place  0.10
2        Playground  0.10
3  Sushi Restaurant  0.10
4              Bank  0.10


----Laurelton----
                  venue  freq
0  Caribbean Restaurant  0.33
1                  Park  0.17
2          Liquor Store  0.17
3      Sculpture Garden  0.17
4         Train Station  0.17


----Lefrak City----
                venue  freq
0    Department Store  0.11
1              Bakery  0.07
2      Cosmetics Shop  0.07
3                 Spa  0.04
4  Frozen Yogurt Shop  0.04


----Lindenwood----
                venue  freq
0       Deli / Bodega  0.08
1         Pizza Place  0.08
2  Chinese Restaurant  0.08
3                 Gym  0.08
4                Park  0.08


----Little Neck----
                venue  freq
0  Chinese Restaura

In [67]:
#Function to sort venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

# Top 10 venues for each neighborhood

In [68]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = queens_grouped['Neighborhood']

for ind in np.arange(queens_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(queens_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood            1st Most Common Venue  \
0                Arverne                        Surf Spot   
1                Astoria                              Bar   
2        Astoria Heights                       Playground   
3             Auburndale               Italian Restaurant   
4            Bay Terrace                   Clothing Store   
5                Bayside                              Bar   
6              Bayswater                     Tennis Court   
7             Beechhurst                              Gym   
8               Bellaire                    Deli / Bodega   
9           Belle Harbor                            Beach   
10             Bellerose                      Pizza Place   
11            Blissville                            Hotel   
12          Breezy Point                       Board Shop   
13             Briarwood                       Playground   
14         Broad Channel                         Dive Bar   
15            Brookville                    Deli / Bodega   
16       Cambria Heights             Caribbean Restaurant   
17         College Point                           Bakery   
18                Corona               Mexican Restaurant   
19            Douglaston               Italian Restaurant   
20         East Elmhurst                       Donut Shop   
21              Edgemere                    Metro Station   
22              Elmhurst               Mexican Restaurant   
23          Far Rockaway                      Pizza Place   
24           Floral Park                Indian Restaurant   
25              Flushing               Chinese Restaurant   
26          Forest Hills             Gym / Fitness Center   
27  Forest Hills Gardens                           Bakery   
28         Fresh Meadows                      Bus Station   
29             Glen Oaks                         Pharmacy   
30              Glendale              Arts & Crafts Store   
31               Hammels                            Beach   
32             Hillcrest                    Grocery Store   
33                Hollis                   Baseball Field   
34            Holliswood                Mobile Phone Shop   
35          Howard Beach               Italian Restaurant   
36         Hunters Point               Italian Restaurant   
37       Jackson Heights        Latin American Restaurant   
38        Jamaica Center                Mobile Phone Shop   
39       Jamaica Estates                Indian Restaurant   
40         Jamaica Hills                         Pharmacy   
41           Kew Gardens               Chinese Restaurant   
42     Kew Gardens Hills                      Bus Station   
43             Laurelton             Caribbean Restaurant   
44           Lefrak City                 Department Store   
45            Lindenwood                            Hotel   
46           Little Neck               Chinese Restaurant   
47      Long Island City                      Coffee Shop   
48                 Malba                     Tennis Court   
49               Maspeth                      Pizza Place   
50        Middle Village                           Bakery   
51           Murray Hill                Korean Restaurant   
52              Neponsit                            Beach   
53          North Corona                    Deli / Bodega   
54       Oakland Gardens                Korean Restaurant   
55            Ozone Park                         Pharmacy   
56               Pomonok                       Playground   
57        Queens Village                             Bank   
58       Queensboro Hill               Chinese Restaurant   
59          Queensbridge                            Hotel   
60            Ravenswood               Italian Restaurant   
61             Rego Park                           Bakery   
62         Richmond Hill                      Pizza Place   
63             Ridgewood                    Deli / Bodega   
64              Rochdale  Southern / Soul Food 

In [66]:
neighborhoods_venues_sorted.iloc[47,]

Neighborhood                  Long Island City
1st Most Common Venue              Coffee Shop
2nd Most Common Venue                    Hotel
3rd Most Common Venue                     Café
4th Most Common Venue       Mexican Restaurant
5th Most Common Venue              Pizza Place
6th Most Common Venue                      Bar
7th Most Common Venue     Gym / Fitness Center
8th Most Common Venue            Deli / Bodega
9th Most Common Venue                      Gym
10th Most Common Venue              Donut Shop
Name: 47, dtype: object

# Cluster the Queens Borough using K-Means

In [69]:
# set number of clusters
kclusters = 5

queens_grouped_clustering = queens_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=2).fit(queens_grouped_clustering)

# check cluster labels generated for each row in the dataframe
#kmeans.labels_[0:10] 
kmeans.labels_

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

# Dataframe that includes the cluster of each neighborhood

In [70]:
queens_merged = queens_data

# add clustering labels
queens_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
queens_merged = queens_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

queens_merged.head() # check the last columns!

Borough     Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Queens          Astoria  40.768509 -73.915654               0   
1  Queens         Woodside  40.746349 -73.901842               0   
2  Queens  Jackson Heights  40.751981 -73.882821               0   
3  Queens         Elmhurst  40.744049 -73.881656               0   
4  Queens     Howard Beach  40.654225 -73.838138               0   

       1st Most Common Venue      2nd Most Common Venue 3rd Most Common Venue  \
0                        Bar  Middle Eastern Restaurant            Hookah Bar   
1              Grocery Store            Thai Restaurant                Bakery   
2  Latin American Restaurant  South American Restaurant   Peruvian Restaurant   
3         Mexican Restaurant            Thai Restaurant    Chinese Restaurant   
4         Italian Restaurant                   Pharmacy            Bagel Shop   

   4th Most Common Venue      5th Most Common Venue 6th Most Common Venue  \
0       Greek Restaurant         Seafood Restaurant                Bakery   
1                    Pub                 Restaurant            Donut Shop   
2                 Bakery         Mexican Restaurant     Mobile Phone Shop   
3  Vietnamese Restaurant  South American Restaurant       Bubble Tea Shop   
4         Sandwich Place                        Spa  Fast Food Restaurant   

   7th Most Common Venue 8th Most Common Venue      9th Most Common Venue  \
0     Italian Restaurant           Pizza Place              Deli / Bodega   
1    Filipino Restaurant                   Bar  Latin American Restaurant   
2     Spanish Restaurant       Thai Restaurant                 Shoe Store   
3  Indonesian Restaurant     Hotpot Restaurant                       Bank   
4     Mexican Restaurant           Flower Shop              Deli / Bodega   

  10th Most Common Venue  
0      Indian Restaurant  
1    American Restaurant  
2             Kids Store  
3   Colombian Restaurant  
4     Chinese Restaurant

In [71]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(queens_merged['Latitude'], queens_merged['Longitude'], queens_merged['Neighborhood'], queens_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examine Queens Cluster Neighborhood

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster.

In [70]:
queens_cluster_0 = queens_merged.loc[queens_merged['Cluster Labels'] == 0, queens_merged.columns[[1] + list(range(4, queens_merged.shape[1]))]]

queens_cluster_1 = queens_merged.loc[queens_merged['Cluster Labels'] == 1, queens_merged.columns[[1] + list(range(4, queens_merged.shape[1]))]]

queens_cluster_2 = queens_merged.loc[queens_merged['Cluster Labels'] == 2, queens_merged.columns[[1] + list(range(4, queens_merged.shape[1]))]]

queens_cluster_3 = queens_merged.loc[queens_merged['Cluster Labels'] == 3, queens_merged.columns[[1] + list(range(4, queens_merged.shape[1]))]]

queens_cluster_4 = queens_merged.loc[queens_merged['Cluster Labels'] == 4, queens_merged.columns[[1] + list(range(4, queens_merged.shape[1]))]]

In [71]:

queens_cluster_0

Neighborhood  Cluster Labels            1st Most Common Venue  \
0                Astoria               0                              Bar   
1               Woodside               0                    Grocery Store   
2        Jackson Heights               0        Latin American Restaurant   
3               Elmhurst               0               Mexican Restaurant   
4           Howard Beach               0               Italian Restaurant   
5                 Corona               0               Mexican Restaurant   
6           Forest Hills               0             Gym / Fitness Center   
7            Kew Gardens               0               Chinese Restaurant   
8          Richmond Hill               0                      Pizza Place   
9               Flushing               0               Chinese Restaurant   
10      Long Island City               0                      Coffee Shop   
11             Sunnyside               0                      Pizza Place   
12         East Elmhurst               0                       Donut Shop   
13               Maspeth               0                      Pizza Place   
14             Ridgewood               0                    Deli / Bodega   
16             Rego Park               0                           Bakery   
17             Woodhaven               0                    Deli / Bodega   
18            Ozone Park               0                         Pharmacy   
19      South Ozone Park               0                    Deli / Bodega   
20         College Point               0                           Bakery   
21            Whitestone               0                     Dance Studio   
22               Bayside               0                              Bar   
23            Auburndale               0               Italian Restaurant   
24           Little Neck               0               Chinese Restaurant   
25            Douglaston               0               Italian Restaurant   
26             Glen Oaks               0                         Pharmacy   
27             Bellerose               0                      Pizza Place   
28     Kew Gardens Hills               0                      Bus Station   
29         Fresh Meadows               0                      Bus Station   
30             Briarwood               0                       Playground   
31        Jamaica Center               0                Mobile Phone Shop   
32       Oakland Gardens               0                Korean Restaurant   
33        Queens Village               0                             Bank   
35         South Jamaica               0                      Bus Station   
36            St. Albans               0             Caribbean Restaurant   
37              Rochdale               0  Southern / Soul Food Restaurant   
38   Springfield Gardens               0                      Bus Station   
39       Cambria Heights               0             Caribbean Restaurant   
40              Rosedale               0                      Bus Station   
41          Far Rockaway               0                      Pizza Place   
42         Broad Channel               0                         Dive Bar   
43          Breezy Point               0                       Board Shop   
44              Steinway               0                    Deli / Bodega   
45            Beechhurst               0                              Gym   
46           Bay Terrace               0                   Clothing Store   
47              Edgemere               0                    Metro Station   
48               Arverne               0                        Surf Spot   
49        Rockaway Beach               0                            Beach   
50              Neponsit               0                            Beach   
51           Murray Hill               0                Korean Restaurant   
53            Holliswood               0                Mobile Phone Shop   
54      

In [72]:

queens_cluster_1

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
15     Glendale               1   Arts & Crafts Store               Brewery   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
15    Chinese Restaurant     Food & Drink Shop           Fish Market   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
15                  Farm        Farmers Market  Fast Food Restaurant   

   9th Most Common Venue 10th Most Common Venue  
15   Filipino Restaurant      Fish & Chips Shop

In [73]:
queens_cluster_2

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
34       Hollis               2        Baseball Field                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
34        Sandwich Place   Fried Chicken Joint         Shopping Mall   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
34    Chinese Restaurant          Gourmet Shop                Bakery   

   9th Most Common Venue 10th Most Common Venue  
34         Grocery Store                 Lounge

In [74]:
queens_cluster_3

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
52  Floral Park               3     Indian Restaurant            Nail Salon   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
52         Grocery Store           Pizza Place    Salon / Barbershop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
52      Basketball Court           Yoga Studio   Empanada Restaurant   

   9th Most Common Venue 10th Most Common Venue  
52           Event Space     Falafel Restaurant

In [75]:
queens_cluster_4

Neighborhood  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
69       Utopia               4         Deli / Bodega   Rental Car Location   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
69                   Spa            Playground    Chinese Restaurant   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
69  South American Restaurant            Donut Shop                Bakery   

   9th Most Common Venue 10th Most Common Venue  
69       Automotive Shop       Basketball Court

# Discussion

The geographical coordinate of Toronto, Canada
are 43.7170226, -79.4197830350134.Scarborough has 81 distinct venues in 54 categories
The most common venues in Scarborough are coffee
shops, international restaurants, and Pharmacy.

Queens has 1805 distinct venues in 269 categories
The most common venues in Queens are bodegas,bakeries, coffee shops, and fitness centre. 
The geographical coordinate of New York City are 40.7308619, -73.9871558. 

Many of the neighborhoods are homogenous and are very similar to each other. Both Scarborough and Queens borough consist of neighborhood cluster that contain majority of the neighborhoods, and the remaining cluster had 1-5 neighborhoods. Queens borough had a significant more number of neighborhoods and venues than Scarborough. 

# Conclusion
In conclusion, based on the quantity of venues and variety of venues, I would choose Queens over
Scarborough as a choice to relocate to Queens offer way more in choices for restaurants, gyms, grocery stores, and extracurricular activities for individuals and families of the company’s employees.